tests/fault_tolerance/deploy/client.py

```diff
@@ -29,194 +30,499 @@
-}
-
-
 # Configure logging
 logging.basicConfig(
     level=logging.INFO,
     format=LOG_FORMAT,
-    datefmt=DATE_FORMAT,  # ISO 8601 UTC format
+    datefmt=DATE_FORMAT,
 )
 
 
-def _get_random_prompt(length):
-    word_list = [f"{i}" for i in range(10)]
-    return " ".join(random.choices(word_list, k=length))
+def get_frontend_port(
+    managed_deployment: ManagedDeployment,
+    client_index: int,
+    deployment_spec: Any,
+    pod_ports: Dict[str, Any],
+    logger: logging.Logger,
+) -> Tuple[Optional[str], Optional[int], Optional[str]]:
+    """
+    Select a frontend pod using round-robin and setup port forwarding.
 
+    Args:
+        managed_deployment: ManagedDeployment instance
+        client_index: Client index for round-robin selection
+        deployment_spec: Deployment specification with port info
+        pod_ports: Dictionary to track existing port forwards
+                  - Key: pod name (str)
+                  - Value: port forward object from managed_deployment.port_forward()
+        logger: Logger instance
 
-def _single_request(
-    url,
-    pod,
-    payload,
-    model,
-    logger,
-    retry_attempts=1,
-    input_token_length=100,
-    output_token_length=100,
-    timeout=30,
-    retry_delay=1,
-):
-    prompt = _get_random_prompt(input_token_length)
-    payload_copy = deepcopy(payload)
-    payload_copy["messages"][0]["content"] = prompt
-    payload_copy["max_tokens"] = output_token_length
-    payload_copy["min_tokens"] = output_token_length
-    payload_copy["model"] = model
-    response = None
-    end_time = None
-    start_time = time.time()
-    results = []
-
-    while retry_attempts:
-        start_request_time = time.time()
-        response = None
+    Returns:
+        Tuple of (pod_name, local_port, pod_instance) or (None, None, None) if failed
+    """
+    pods = managed_deployment.get_pods(managed_deployment.frontend_service_name)
+
+    port = 0
+    pod_name = None
+    selected_pod = None
+
+    # Filter ready pods and cleanup stale port forwards
+    pods_ready = []
+
+    for pod in pods[managed_deployment.frontend_service_name]:
+        if pod.ready():
+            pods_ready.append(pod)
+        else:
+            # Cleanup port forwards for non-ready pods
+            if pod.name in pod_ports:
+                try:
+                    pod_ports[pod.name].stop()
+                except Exception as e:
+                    logger.debug(f"Error stopping port forward for {pod.name}: {e}")
+                del pod_ports[pod.name]
+
+    if not pods_ready:
+        logger.error("No ready frontend pods found")
+        return None, None, None
+
+    # Round-robin selection based on client index
+    selected_pod = pods_ready[client_index % len(pods_ready)]
+    pod_name = selected_pod.name
+
+    # Setup or reuse port forward
+    if pod_name not in pod_ports:
+        # Get port from deployment_spec (default: 8000)
+        port_value = getattr(deployment_spec, "_port", 8000)
```

Comment:
We have properties for these to make public - why are we using getattr ?

https://github.com/ai-dynamo/dynamo/blob/02c822d6b919ececa248755e8fc64c96123a1257/tests/utils/managed_deployment.py#L198